sympy/simplify/tests/test_cse.py

```diff
@@ -288,8 +288,8 @@ def test_issue_4499():
         [(x0, 2*a), (x1, -b), (x2, x0 + x1), (x3, x2 + 1), (x4, sqrt(z)), (x5,
         B(b - 1, x4)), (x6, -x0), (x7, (x4/2)**(x6 + 1)*G(b)*G(x3)), (x8,
         x7*B(x2, x4)), (x9, B(b, x4)), (x10, x7*B(x3, x4))],
-        [(a, a + 1/2, x0, b, x3, x5*x8, x4*x8*x9, x10*x4*x5, x10*x9,
-        1, 0, 1/2, z/2, x1 + 1, b + x6, x6)])
+        [a, a + 1/2, x0, b, x3, x5*x8, x4*x8*x9, x10*x4*x5, x10*x9,
+        1, 0, 1/2, z/2, x1 + 1, b + x6, x6])
```

Comment:
This change looks dangerous. The information that this was originally a tuple is lost. I would probably also break code downstream.

Reply:
@bjodah what do you mean by break the code downstream?

Reply:
Projects relying on SymPy need it to have a stable API.

Reply:
But inconsistent return types of `replacements` and `reduced_exprs` can be harmful and confusing too. Like for instance to access first `reduced_exprs` in above example what should we do : 
`ans[1][0][0]` or `ans[1][0]` without actually printing the output?

Reply:
The intent of cse was not to return the same type as was passed. It really should be list/list as indicated in the docstring. I don't know how the Tuple was allowed to be returned as a Tuple, but that is not right. Tuple is not an Expr. I suppose it *could* be made to act that way, but then there is more work needed to handle all types of Basic input (instead of Expr).

Reply:
I see, should we incrementally expand the scope of `cse` to support Matrices etc.? (or should we expect the selected classes where CSE makes sense to implement their own methods for that?)

Reply:
Perhaps it would be best to poll at the discussion list (mailing list). Like subs, perhaps cse could first see if there is an `_eval_cse` method and if not, raise an error if the input is not an Expr or fails `iterable(foo)`.

Reply:
> But inconsistent return types of replacements and reduced_exprs can be harmful and confusing too. Like for instance to access first element of reduced_exprs in above example, what should we do : ans[1][0][0] or ans[1][0] without actually printing the output?

These are completely different things. `replacements` is always a list of tuples, where the first element of the tuple is a dummy symbol and the second is the expression represented by that symbol. It's the exact format accepted by `subs` to undo the CSE.

Reply:
yeah, `replacements` is always a list of tuples but that's not true for `reduced_exprs` in current master (contradicting [doctring](https://github.com/sympy/sympy/blob/35e1eda72b1d8c693d87f9dca7d1a5775c2d0341/sympy/simplify/cse_main.py#L649)).
But I think @bjodah is right, any changes now are quite dangerous with `cse` being such an important and widely used method.